### PR TITLE
FIX-#2616: Add config for num partitions, deprecate DEFAULT_NPARTITIONS

### LINKED ISSUE #2616: Manual partitions number setting in tests is broken
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04
- **Modin version**: d6637304a24f986b5abb12e2b525bc3af97d6cac
- **Python version**:  3.8.6
- **Code we can use to reproduce**:
```
import os

os.environ["MODIN_ENGINE"] = "ray"

import modin.pandas as pd
pd.DEFAULT_NPARTITIONS = 1

data = [x for x in range(10000)]

df = pd.DataFrame(data)
print(len(df._query_compiler._modin_frame._partitions)) # <available number of cores>
```

### Describe the problem

Occurred after 9a6695d3b36cb6c21e835b10d62f26a474583f15 commit.

### PR DESCRIPTION
**Pull request by devin-petersohn**
* Resolves #2616

Creates a new configuration for setting or updating the number of
partitions. Deprecate DEFAULT_NPARTITIONS in favor of something slightly
more robust.

Signed-off-by: Devin Petersohn <devin.petersohn@gmail.com>

## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/CONTRIBUTING.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2616 
- [x] tests added and passing

### CODE REVIEW

File: modin/test/test_envvar_npartitions.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+@pytest.mark.parametrize("left_num_partitions", [2, 4, 6, 8, 10])
+@pytest.mark.parametrize("right_num_partitions", [2, 4, 6, 8, 10])
+def test_runtime_change_npartitions(left_num_partitions, right_num_partitions):
+    NPartitions.put(left_num_partitions)
+    data = np.random.randint(0, 100, size=(2 ** 16, 2 ** 8))
+    left_df = pd.DataFrame(data)
+    part_shape = left_df._query_compiler._modin_frame._partitions.shape
+    assert part_shape[0] == left_num_partitions and part_shape[1] == min(
+        left_num_partitions, 8
```
Comment (YarShev) [inline modin/test/test_envvar_npartitions.py:38]: Why is the minimum chosen from `left_num_partitions` and `8`, not `left_num_partitions` and `4`, for instance?

Comment (devin-petersohn) [reply modin/test/test_envvar_npartitions.py:38]: Column partitions are currently minimum 32 columns. In this data, that means that there will be at most 8 partitions because there is 2**8 columns.

File: modin/pandas/__init__.py [hunk 2]
```diff
@@ -90,9 +90,8 @@
 
 # Set this so that Pandas doesn't try to multithread by itself
 os.environ["OMP_NUM_THREADS"] = "1"
-DEFAULT_NPARTITIONS = 4
-num_cpus = 1
 
+num_cpus = 1
```
Comment (YarShev) [inline modin/pandas/__init__.py:94]: This should be removed because no used anywhere. Also, please remove it [here](https://github.com/modin-project/modin/blob/master/modin/pandas/test/test_api.py#L65).

File: modin/test/test_envvar_npartitions.py (hunk 1 above)
Comment (YarShev) [reply modin/test/test_envvar_npartitions.py:38]: Got it, thanks!

**Review by YarShev — CHANGES_REQUESTED** [review]
Please, fix remaining `DEFAULT_NPARTITIONS` entries because not all are changed.
Also, why is `modin/test/__init__.py empty?

**Comment by YarShev** [comment]
On new tests, they should be added to CI.

File: modin/pandas/__init__.py (hunk 2 above)
Comment (devin-petersohn) [reply modin/pandas/__init__.py]: There were many uses of it, removed all of them.

**Comment by devin-petersohn** [comment]
>Also, why is `modin/test/__init__.py` empty?

The init was automatically added by Pycharm. It doesn't necessarily need to be there, but we have them in other test locations and they are empty there as well.

**Comment by YarShev** [comment]
> > Also, why is `modin/test/__init__.py` empty?
> 
> The init was automatically added by Pycharm. It doesn't necessarily need to be there, but we have them in other test locations and they are empty there as well.

I mean the license should be added at the top of the file.

**Comment by YarShev** [comment]
Also, there are some more places with `DEFAULT_NPARTITIONS` entries. They should be replaced/removed. Here are the files: modin/docs/troubleshooting.rst, modin/examples/tutorial/tutorial_notebooks/cluster/exercise_4.ipynb, modin/examples/tutorial/tutorial_notebooks/cluster/exercise_5.ipynb, modin/engines/base/io/text/text_file_dispatcher.py, modin/pandas/test/test_api.py.

**Comment by devin-petersohn** [comment]
@YarShev I have created #2635 to fix troubleshooting documentation. It is severely out of date and probably not worth changing in this PR.

**Comment by YarShev** [comment]
`DEFAULT_NPARTITIONS` should be removed [here](https://github.com/modin-project/modin/blob/master/modin/pandas/test/test_api.py#L48) too.

**Comment by YarShev** [comment]
> @YarShev I have created #2635 to fix troubleshooting documentation. It is severely out of date and probably not worth changing in this PR.

That makes sense to fix in another PR then.

File: examples/tutorial/tutorial_notebooks/cluster/exercise_4.ipynb [hunk 3]
```diff
@@ -143,4 +143,4 @@
  },
  "nbformat": 4,
  "nbformat_minor": 2
-}
+}
```
Comment (YarShev) [inline examples/tutorial/tutorial_notebooks/cluster/exercise_4.ipynb:146]: blank line is needed

File: examples/tutorial/tutorial_notebooks/cluster/exercise_5.ipynb [hunk 4]
```diff
@@ -181,4 +182,4 @@
  },
  "nbformat": 4,
  "nbformat_minor": 2
-}
+}
```
Comment (YarShev) [inline examples/tutorial/tutorial_notebooks/cluster/exercise_5.ipynb:185]: blank line is needed

**Comment by YarShev** [comment]
LGTM! Will wait for CI.

**Review by YarShev — APPROVED** [review]
